moshi/src/main/java/com/squareup/moshi/JsonWriter.kt

```diff
@@ -0,0 +1,548 @@
+   * without nesting.
+   *
+   *
+   * For example, the following creates JSON with nested arrays: `[1,[2,3,4],5]`.
+   *
+   * ```java
+   * JsonAdapter<List<Integer>> integersAdapter = ...
+   * public void writeNumbers(JsonWriter writer) {
+   *   writer.beginArray();
+   *   writer.value(1);
+   *   integersAdapter.toJson(writer, Arrays.asList(2, 3, 4));
+   *   writer.value(5);
+   *   writer.endArray();
+   * }
+   * ```
+   *
+   *
+   * With flattening we can create JSON with a single array `[1,2,3,4,5]`:
+   *
+   * ```java
+   * JsonAdapter<List<Integer>> integersAdapter = ...
+   *
+   * public void writeNumbers(JsonWriter writer) {
+   *   writer.beginArray();
+   *   int token = writer.beginFlatten();
+   *   writer.value(1);
+   *   integersAdapter.toJson(writer, Arrays.asList(2, 3, 4));
+   *   writer.value(5);
+   *   writer.endFlatten(token);
+   *   writer.endArray();
+   * }
+   * ```
+   *
+   *
+   * This method flattens arrays within arrays:
+   *
+   * Emit:       `[1, [2, 3, 4], 5]`
+   * To produce: `[1, 2, 3, 4, 5]`
+   *
+   * It also flattens objects within objects. Do not call [name] before writing a flattened
+   * object.
+   *
+   * Emit:       `{"a": 1, {"b": 2}, "c": 3}`
+   * To Produce: `{"a": 1, "b": 2, "c": 3}`
+   *
+   *
+   * Other combinations are permitted but do not perform flattening. For example, objects inside of
+   * arrays are not flattened:
+   *
+   * Emit:      ` [1, {"b": 2}, 3, [4, 5], 6]`
+   * To Produce: `[1, {"b": 2}, 3, 4, 5, 6]`
+   *
+   *
+   * This method returns an opaque token. Callers must match all calls to this method with a call
+   * to [endFlatten] with the matching token.
+   */
+  @CheckReturnValue
+  public fun beginFlatten(): Int {
+    val context = peekScope()
+    check(!(context != JsonScope.NONEMPTY_OBJECT && context != JsonScope.EMPTY_OBJECT && context != JsonScope.NONEMPTY_ARRAY && context != JsonScope.EMPTY_ARRAY)) { "Nesting problem." }
+    val token = flattenStackSize
+    flattenStackSize = stackSize
+    return token
+  }
+
+  /** Ends nested call flattening created by [beginFlatten].  */
+  public fun endFlatten(token: Int) {
+    flattenStackSize = token
+  }
+
+  /**
+   * Returns a [JsonPath](http://goessner.net/articles/JsonPath/) to the current location
+   * in the JSON value.
+   */
+  @get:CheckReturnValue
+  public val path: String
+    get() = JsonScope.getPath(stackSize, scopes, pathNames, pathIndices)
+
+  /** Returns the tag value for the given class key.  */
+  @CheckReturnValue
+  public fun <T> tag(clazz: Class<T>): T? {
+    @Suppress("UNCHECKED_CAST")
+    return if (tags == null) null else tags!![clazz] as T?
+  }
+
+  /** Assigns the tag value using the given class key and value.  */
+  public fun <T> setTag(clazz: Class<T>, value: T) {
+    require(clazz.isAssignableFrom(value!!::class.java)) { "Tag value must be of type " + clazz.name }
+    if (tags == null) {
+      tags = LinkedHashMap()
+    }
+    tags!![clazz] = value
```

Comment:
Why the `!!` here?

Reply:
Smart cast to 'MutableMap<Class<*>, Any>' is impossible, because 'tags' is a mutable property that could have been changed by this time

Reply:
Yes, you will need to write it differently

Reply:
This remains unresolved

Reply:
I'm not sure what the "differently" you mean.